### PR TITLE
Rename #expires_now and add #http_cache_never

### PR DESCRIPTION
**Pull request by nateberkopec**
See https://github.com/rails/rails/pull/19556. cc @sgrif who was assigned to that PR.

### CODE REVIEW

**Comment by nateberkopec** [comment]
This method is still very poorly named, but I'm not pr0 enough to know how Rails might handle deprecating/changing the name of such an old method.

**Comment by pixeltrix** [comment]
@nateberkopec deprecation of an existing method is straightforward - just wrap a call to the new method from the old method with a warning, e.g:

``` ruby
def old_method(*args)
  ActiveSupport::Deprecation.warn("old_method is deprecated and will be removed in Rails 5.1. Please use new method instead")
  new_method(*args)
end
```

Then in Rails 5.1 we just delete the old method. :smile:

**Comment by nateberkopec** [comment]
@pixeltrix Cool, updated with what I think we should actually do. Commit message:

>The original documentation on expires_now was slightly misleading, as in most
contexts "cache" is the same thing as "store". In HTTP caching, this is not
the case, as something which is "not cached" may still be stored and revalidated
on each request. Some users may have though the expires_now method prevented
browsers and intermediate caches from storing responses, which it would not.
>
> In addition, expires_now was a bit of a confusing name for adding a cache-control
directive of "no-cache". By renaming "expires_now" to "always_revalidate", we are
revealing the intention behind the method.
>
> Also, provided is an "http_cache_never" method intended to be the anti-counterpart
to "http_cache_forever". It adds a "no-store" directive, which will prevent the
response from being stored *or* cached by any caches.

**Comment by maclover7** [comment]
Can you add some `assert_deprecated` tests?

File: actionpack/lib/action_controller/metal/conditional_get.rb [hunk 1]
```diff
@@ -208,12 +208,31 @@ def expires_in(seconds, options = {})
...
+      always_revalidate(*args)
+    end
+
+    # Sets a HTTP 1.1 Cache-Control header of <tt>no-cache</tt>. Browsers and
+    # intermediate caches will still store the response, but will always
+    # revalidate to see if the response has changed. Use for frequently-updated
+    # resources that need to be revalidated on each request, but may sometimes
+    # be unchanged.
+    def always_revalidate
       response.cache_control.replace(:no_cache => true)
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/conditional_get.rb]: new hash syntax

File: actionpack/lib/action_controller/metal/conditional_get.rb [hunk 2]
```diff
@@ -208,12 +208,31 @@ def expires_in(seconds, options = {})
...
     end
 
+    # Sets a HTTP 1.1 Cache-Control header of <tt>no-store</tt>. Browsers and
+    # intermediate caches will never store this response and will never cache it.
+    # Use for sensitive or private information, but be aware that intermediates
+    # or eavesdroppers may ignore this directive and privacy cannot be guaranteed.
+    def http_cache_never
+      response.cache_control.replace(:no_store => true)
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/conditional_get.rb]: New hash syntax

File: actionpack/test/controller/render_test.rb [hunk 3]
```diff
@@ -113,14 +113,19 @@ def conditional_hello_with_expires_in_with_public_with_more_keys_old_syntax
...
+  def conditional_hello_with_always_revalidate
+    always_revalidate
     render :action => 'hello_world'
   end
 
   def conditional_hello_with_cache_control_headers
     response.headers['Cache-Control'] = 'no-transform'
-    expires_now
+    always_revalidate
+    render :action => 'hello_world'
```
Comment (rafaelfranca) [inline actionpack/test/controller/render_test.rb]: new hash syntax

**Comment by nateberkopec** [comment]
Fixed hash syntax and added deprecation test.

I'm confused on why my tests are failing though - something is replace the `no-store` directive with `private`, which doesn't even make sense, as they are not the same thing. Can someone give me a lead as to what's doing this?

Also changed the behavior of `always_revalidate` to add a directive rather than replace all of them. This is because it may be useful to specify a header of `private no-cache` - only allow browsers to store this response, and they must always revalidate it.

Pinging @matthewd who seems to ~~also be a sadist~~ enjoy reading W3C specs as much as I do, let me know what you think about this change.

**Comment by nateberkopec** [comment]
Also, I'm wondering if anyone thinks that `always_revalidate` is sort of a silly method now - you can just pass `no_cache` as an option to `expires_in`.

**Pull request closed without merging**